keps/sig-scheduling/20200612-simplified-topology-manager.md

```diff
@@ -0,0 +1,229 @@
+
+## Plugin implementation details
+
+Since topology of the node is stored in the CRD, kube-scheduler should be subscribed for updates of appropriate CRD type. Kube-scheduler will use informers which will be generated with the name NodeTopologyInformer. NodeTopologyInformer will run in TopologyMatch plugin.
+
+### Topology information in the TopologyMatch plugin
+
+Once NodeResourceTopology is received TopologyMatch plugin keeps it in its own state of type
+NodeTopologyMap. This state will be used every time when scheduler needs to make a decidion based on node topology.
+
+```go
+
+type NodeTopologyMap map[string]topologyv1alpha1.NodeResourceTopology
+
+type NodeResourceTopology struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    TopologyPolicy string        `json:"topologyPolicy"`
+    Nodes   []NUMANodeResource   `json:"nodes"`
+}
+
+// NUMANodeResource is the spec for a NodeResourceTopology resource
+type NUMANodeResource struct {
+    NUMAID int
+    Resources v1.ResourceList
+}
+```
+Where TopologyPolicy may have following values: none, best-effort, restricted, single-numa-node
+
+To use these labels both in kube-scheduler and in kubelet string constants of these labels should be moved from pkg/kubelet/cm/topologymanager/ and pkg/kubelet/apis/config/types.go to pkg/apis/core/types.go a one single place.
+
+NUMAID is an auxiliary field since scheduler version of TopologyManager doesn't make a real assignment.
+
+### Description of the Algorithm
+
+The algorithm which implements single-numa-node policy is following:
+
+```go
+if !guaranteedQoS(pod) {
+    return nil
+}
+for _, container := range(containers) {
+    bitmask := bm.NewEmptyBitMask()
+    for resource, quantity := range(container.Resources.Requests) {
+        resourceBitmask := bm.NewEmptyBitMask()
+        for _, numaNode := range(nodes) {
+            numaQuantity, ok := numaNode.Resources[resource]
+            if !ok || numaQuantity.Cmp(quantity) < 0 {
+                continue
+            }
+            resourceBitmask.Add(numaNode.NUMAID)
+        }
+    }
+    if resourceBitmask.IsEmpty() {
+        continue
+    }
+    bitmask.And(resourceBitmask)
+}
+if bitmask.IsEmpty() {
+    // definitely we can't align container, so we can't align a pod
+    return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Can't align container: %s", container.Name))
+}
+```
+## Accessing NodeResourceTopology CRD
+
+In order to allow the scheduler (deployed as a pod) to access NodeResourceTopology CRD instances, ClusterRole and ClusterRoleBinding would have to be configured as below:
+
+``` yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: noderesourcetopology-handler
+rules:
+- apiGroups: ["k8s.cncf.io"]
```

Comment:
where is this API group being defined?

i would have expected something in the x-k8s.io namespace if this was being done in the kubernetes project.

Reply:
API definition in captured in KEP: #1870. Sure, we can update it to something like "topology.node.kubernetes.io" or  node.k8s.io"